File: iOS/GlobusCRM/GlobusCRM/Features/Health/Data/Remote/DTOs/HealthDTOs.swift
import Foundation

// GET /wellness/portal/prescriptions — plain array, no wrapper.
// `drugs` may be a JSON string, a real array, or a wrapped object depending on backend environment.
struct PrescriptionDTO: Decodable {
    let id: Int
    let visitId: Int?
    let drugs: String?
    let instructions: String?
    let pdfUrl: String?
    let visit: PrescriptionVisitDTO?
    let doctor: PrescriptionDoctorDTO?
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id, visitId, drugs, instructions, pdfUrl, visit, doctor, createdAt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(Int.self, forKey: .id)
        visitId = try container.decodeIfPresent(Int.self, forKey: .visitId)
        instructions = try container.decodeIfPresent(String.self, forKey: .instructions)
        pdfUrl = try container.decodeIfPresent(String.self, forKey: .pdfUrl)
        visit = try container.decodeIfPresent(PrescriptionVisitDTO.self, forKey: .visit)
        doctor = try container.decodeIfPresent(PrescriptionDoctorDTO.self, forKey: .doctor)
        createdAt = try container.decodeIfPresent(String.self, forKey: .createdAt)

        if let stringValue = try? container.decodeIfPresent(String.self, forKey: .drugs) {
            drugs = stringValue
        } else if let jsonValue = try? container.decodeIfPresent(FlexibleJSONValue.self, forKey: .drugs) {
            drugs = jsonValue.jsonString
        } else {
            drugs = nil
        }
    }
}

struct PrescriptionVisitDTO: Decodable {
    let id: Int?
    let visitDate: String?
    let service: PrescriptionServiceDTO?
}

struct PrescriptionServiceDTO: Decodable {
    let name: String
}

struct PrescriptionDoctorDTO: Decodable {
    let id: Int?
    let name: String?
}

struct DrugDTO: Decodable {
    let id: String?
    let name: String?
    let dosage: String?
    let frequency: String?
    let duration: String?
    let instructions: String?
}

private enum FlexibleJSONValue: Decodable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case object([String: FlexibleJSONValue])
    case array([FlexibleJSONValue])
    case null

    init(from decoder: Decoder) throws {
        if let keyed = try? decoder.container(keyedBy: DynamicCodingKey.self) {
            var object: [String: FlexibleJSONValue] = [:]
            for key in keyed.allKeys {
                object[key.stringValue] = try keyed.decode(FlexibleJSONValue.self, forKey: key)
            }
            self = .object(object)
            return
        }

        if var unkeyed = try? decoder.unkeyedContainer() {
            var array: [FlexibleJSONValue] = []
            while !unkeyed.isAtEnd {
                array.append(try unkeyed.decode(FlexibleJSONValue.self))
            }
            self = .array(array)
            return
        }

        let single = try decoder.singleValueContainer()
        if single.decodeNil() {
            self = .null
        } else if let value = try? single.decode(String.self) {
            self = .string(value)
        } else if let value = try? single.decode(Double.self) {
            self = .number(value)
        } else if let value = try? single.decode(Bool.self) {
            self = .bool(value)
        } else {
            self = .null
        }
    }

    var jsonString: String? {
        if case .string(let value) = self {
            return value
        }

        guard let object = jsonObject,
              JSONSerialization.isValidJSONObject(object),
              let data = try? JSONSerialization.data(withJSONObject: object),
              let string = String(data: data, encoding: .utf8) else {
            return nil
        }
        return string
    }

    private var jsonObject: Any? {
        switch self {
        case .string(let value):
            return value
        case .number(let value):
            return value
        case .bool(let value):
            return value
        case .object(let value):
            return value.compactMapValues { $0.jsonObject }
        case .array(let value):
            return value.compactMap { $0.jsonObject }
        case .null:
            return NSNull()
        }
    }
}

private struct DynamicCodingKey: CodingKey {
    let stringValue: String
    let intValue: Int?

    init?(stringValue: String) {
        self.stringValue = stringValue
        self.intValue = nil
    }

    init?(intValue: Int) {
        self.stringValue = String(intValue)
        self.intValue = intValue
    }
}

// GET /wellness/patients/{patientId}/treatment-plans — plain array, no wrapper
// Real field names confirmed from Android staging 2026-06-04
struct TreatmentPlanDTO: Decodable {
    let id: Int
    let name: String
    let totalSessions: Int
    let completedSessions: Int
    let startedAt: String?
    let nextDueAt: String?
    let status: String
    let totalPrice: Double?
    let service: TreatmentServiceDTO?
}

struct TreatmentServiceDTO: Decodable {
    let id: Int?
    let name: String
    let category: String?
}

// GET /wellness/patients/{patientId}/consents — plain array, no wrapper
// Real field names confirmed from Android staging 2026-06-04
struct ConsentFormDTO: Decodable {
    let id: Int
    let templateName: String
    let signedAt: String?
    let hasPdfBlob: Bool?
    let service: ConsentServiceDTO?
}

struct ConsentServiceDTO: Decodable {
    let id: Int?
    let name: String
}

File: android/app/src/main/java/com/globus/crm/feature/health/data/remote/dto/HealthDtos.kt
package com.globus.crm.feature.health.data.remote.dto

import com.squareup.moshi.JsonClass

// GET /portal/prescriptions — backend stores drugs as a JSON-encoded string, not an array.
// visit/doctor are nested objects returned by the API.
@JsonClass(generateAdapter = true)
data class PrescriptionDto(
    val id: Int,
    val visitId: Int?,
    val drugs: String,
    val instructions: String?,
    val pdfUrl: String?,
    val visit: PrescriptionVisitDto?,
    val doctor: PrescriptionDoctorDto?,
    val createdAt: String?,
)

@JsonClass(generateAdapter = true)
data class PrescriptionVisitDto(
    val id: Int,
    val visitDate: String?,
    val service: PrescriptionServiceDto?,
)

@JsonClass(generateAdapter = true)
data class PrescriptionServiceDto(
    val name: String,
)

@JsonClass(generateAdapter = true)
data class PrescriptionDoctorDto(
    val id: Int,
    val name: String?,
)

data class DrugDto(
    val name: String,
    val dosage: String?,
    val frequency: String?,
    val duration: String?,
    val instructions: String?,
)

// ── Prescription renewal / medicine requests ─────────────────────────────
// Data layer ready ahead of the UI, same as Treatment Plans / Consents below.
//
// POST /portal/prescription-requests — raise a renewal against one of the
// patient's OWN prescriptions.
//
// `medicines` is the ONLY optional part that changes the meaning of the
// request: omit it (or send an empty list) and the backend treats it as
// "renew the complete prescription". When present, every name must match a
// medicine already on that prescription — the server re-reads the Rx and
// rejects anything else with 400 MEDICINE_NOT_ON_PRESCRIPTION.
//
// Do NOT send a doctor id: the backend resolves the prescriber from the
// prescription itself and ignores any client claim.
@JsonClass(generateAdapter = true)
data class CreatePrescriptionRequestDto(
    val prescriptionId: Int,
    // Medicine names as shown on the prescription; null/empty = whole Rx.
    val medicines: List<String>? = null,
    // "another 2 months" → 60. Server caps at 365.
    val durationDays: Int? = null,
    // Optional explicit window, YYYY-MM-DD.
    val from: String? = null,
    val to: String? = null,
    val notes: String? = null,
)

// Response of POST /portal/prescription-requests and each item of
// GET /portal/prescription-requests.
//
// `isFullPrescription` is the server's own answer to "did this ask for
// everything?" — prefer it over inspecting requestedDrugs, so the app and the
// clinic admin panel never disagree about what a null list meant.
@JsonClass(generateAdapter = true)
data class PrescriptionRequestDto(
    val id: Int,
    val status: String, // PENDING | ACCEPTED | REJECTED | COMPLETED
    val prescriptionId: Int,
    val patientId: Int,
    val doctorId: Int?,
    val doctorName: String?,
    val isFullPrescription: Boolean,
    val requestedDrugs: List<RequestedDrugDto>?,
    val requestedDurationDays: Int?,
    val requestedFrom: String?,
    val requestedTo: String?,
    val notes: String?,
    // Staff's note on the decision — populated on reject / complete.
    val reviewNote: String?,
    val reviewedAt: String?,
    val createdAt: String?,
    val updatedAt: String?,
)

@JsonClass(generateAdapter = true)
data class RequestedDrugDto(
    val name: String?,
    val dosage: String?,
    val frequency: String?,
    val duration: String?,
)

// GET /api/wellness/patients/{patientId}/treatment-plans — CUSTOMER JWT (verifyToken).
// Real shape confirmed against staging 2026-06-04.
@JsonClass(generateAdapter = true)
data class TreatmentPlanDto(
    val id: Int,
    val name: String,
    val totalSessions: Int,
    val completedSessions: Int,
    val startedAt: String,
    val nextDueAt: String?,
    val status: String,
    val totalPrice: Double,
    val patientId: Int,
    val serviceId: Int,
    val tenantId: Int,
    val patient: TreatmentPatientRefDto?,
    val service: TreatmentServiceRefDto?,
)

@JsonClass(generateAdapter = true)
data class TreatmentPatientRefDto(
    val id: Int,
    val name: String,
    val phone: String?,
)

@JsonClass(generateAdapter = true)
data class TreatmentServiceRefDto(
    val id: Int,
    val name: String,
    val category: String?,
)

// GET /api/wellness/patients/{patientId}/consents — CUSTOMER JWT (verifyToken).
// Real shape confirmed against staging 2026-06-04.
@JsonClass(generateAdapter = true)
data class ConsentFormDto(
    val id: Int,
    val templateName: String,
    val signedAt: String,
    val patientId: Int,
    val serviceId: Int,
    val hasPdfBlob: Boolean,
    val patient: ConsentPatientRefDto?,
    val service: ConsentServiceRefDto?,
)

@JsonClass(generateAdapter = true)
data class ConsentPatientRefDto(
    val id: Int,
    val name: String,
)

@JsonClass(generateAdapter = true)
data class ConsentServiceRefDto(
    val id: Int,
    val name: String,
)
